oneclick.py

```diff
@@ -0,0 +1,600 @@
+                        content=description,
+                        media_type="text/html",
+                    )
+                )
+
+            metadata.title = title
+            metadata.language = language
+            metadata.medium = medium
+            metadata.series = series_name
+            metadata.publisher = publisher
+            metadata.published = published
+            metadata.identifiers = identifiers
+            metadata.subjects = subjects
+            metadata.contributors = contributors
+            metadata.links = links
+
+        if include_formats:
+            formats = []
+            if metadata.medium == Edition.BOOK_MEDIUM:
+                content_type, drm_scheme = cls.oneclick_formats.get("ebook-epub-oneclick")
+                formats.append(FormatData(content_type, drm_scheme))
+            elif metadata.medium == Edition.AUDIO_MEDIUM:
+                content_type, drm_scheme = cls.oneclick_formats.get("audiobook-mp3-oneclick")
+                formats.append(FormatData(content_type, drm_scheme))
+            else:
+                cls.log.warn("Unfamiliar format: %s", format_id)
+
+            # Make a CirculationData so we can write the formats, 
+            circulationdata = CirculationData(
+                data_source=DataSource.ONECLICK,
+                primary_identifier=primary_identifier,
+                formats=formats,
+            )
+
+            metadata.circulation = circulationdata
+
+        return metadata
+
+
+
+class OneClickBibliographicCoverageProvider(BibliographicCoverageProvider):
+    """Fill in bibliographic metadata for OneClick records."""
+
+    def __init__(self, _db, input_identifier_types=None, 
+                 metadata_replacement_policy=None, oneclick_api=None,
+                 **kwargs):
+        # We ignore the value of input_identifier_types, but it's
+        # passed in by RunCoverageProviderScript, so we accept it as
+        # part of the signature.
+        
+        oneclick_api = oneclick_api or OneClickAPI(_db)
+        super(OneClickBibliographicCoverageProvider, self).__init__(
+            _db, oneclick_api, DataSource.ONECLICK,
+            batch_size=25, 
+            metadata_replacement_policy=metadata_replacement_policy,
+            **kwargs
+        )
+
+
+    def process_item(self, identifier):
+        """ OneClick availability information is served separately from 
+        the book's metadata.  Furthermore, the metadata returned by the 
+        "book by isbn" request is less comprehensive than the data returned 
+        by the "search titles/genres/etc." endpoint.
+
+        This method hits the "by isbn" endpoint and updates the bibliographic 
+        metadata returned by it. 
+        """
+        try:
+            response_dictionary = self.api.get_metadata_by_isbn(identifier)
+        except ValueError as error:
+            return CoverageFailure(identifier, error.message, data_source=self.output_source, transient=False)
```

Comment:
At the moment, a non-transient error from a bibliographic coverage provider almost never happens. Even if a book isn't in the library's collection (to pick an example of one way you use ValueError), it might become part of the collection later on. I would err on the side of making these errors transient.

In the future we'll resolve this by replacing the transient/permanent field with a 'retry after' field. If a book isn't in the collection the 'retry after' value would be, like, a week.